// cypress-xray-plugin/src/repository/jira/fields/jiraFieldRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachingJiraFieldRepository = void 0;
const dedent_1 = require("../../../util/dedent");
const pretty_1 = require("../../../util/pretty");
const jiraIssueFetcher_1 = require("./jiraIssueFetcher");
/**
 * A Jira field repository which caches retrieved field IDs. After the first ID retrieval, all
 * subsequent accesses will return the cached value.
 */
class CachingJiraFieldRepository {
    jiraClient;
    names = {};
    ids = {};
    /**
     * Constructs a new caching Jira field repository. The Jira client is necessary for accessing
     * Jira data.
     *
     * @param jiraClient - the Jira client
     */
    constructor(jiraClient) {
        this.jiraClient = jiraClient;
    }
    async getFieldId(fieldName) {
        // Lowercase everything to work around case sensitivities.
        // Jira sometimes returns field names capitalized, sometimes it doesn't (?).
        const lowerCasedName = fieldName.toLowerCase();
        if (!(lowerCasedName in this.ids)) {
            const jiraFields = await this.jiraClient.getFields();
            if (!jiraFields) {
                throw new Error(`Failed to fetch Jira field ID for field with name: ${fieldName}`);
            }
            else {
                const matches = jiraFields.filter((field) => {
                    const lowerCasedField = field.name.toLowerCase();
                    return lowerCasedField === lowerCasedName;
                });
                if (matches.length > 1) {
                    throw this.multipleFieldsError(fieldName, matches);
                }
                jiraFields.forEach((jiraField) => {
                    this.ids[jiraField.name.toLowerCase()] = jiraField.id;
                    this.names[jiraField.name.toLowerCase()] = jiraField.name;
                });
            }
        }
        if (!(lowerCasedName in this.ids)) {
            throw this.missingFieldsError(fieldName);
        }
        return this.ids[lowerCasedName];
    }
    multipleFieldsError(fieldName, matches) {
        const nameDuplicates = (0, pretty_1.prettyPadObjects)(matches)
            .map((duplicate) => Object.entries(duplicate)
            .map((entry) => `${entry[0]}: ${entry[1]}`)
            .join(", "))
            .sort()
            .join("\n");
        const idSuggestions = matches.map((field) => `"${field.id}"`).join(" or ");
        return new Error((0, dedent_1.dedent)(`
                Failed to fetch Jira field ID for field with name: ${fieldName}
                There are multiple fields with this name

                Duplicates:
                  ${nameDuplicates}

                You can provide field IDs in the options:

                  jira: {
                    fields: {
                      ${this.getOptionName(fieldName)}: // ${idSuggestions}
                    }
                  }
            `));
    }
    missingFieldsError(fieldName) {
        if (Object.keys(this.ids).length === 0) {
            throw new Error((0, dedent_1.dedent)(`
                    Failed to fetch Jira field ID for field with name: ${fieldName}
                    Make sure the field actually exists and that your Jira language settings did not modify the field's name

                    You can provide field IDs directly without relying on language settings:

                      jira: {
                        fields: {
                          ${this.getOptionName(fieldName)}: // corresponding field ID
                        }
                      }
                `));
        }
        else {
            const availableFields = Object.entries((0, pretty_1.prettyPadValues)(this.names))
                .map((entry) => `name: ${entry[1]} id: ${JSON.stringify(entry[0])}`)
                .sort();
            throw new Error((0, dedent_1.dedent)(`
                    Failed to fetch Jira field ID for field with name: ${fieldName}
                    Make sure the field actually exists and that your Jira language settings did not modify the field's name

                    Available fields:
                      ${availableFields.join("\n")}

                    You can provide field IDs directly without relying on language settings:

                      jira: {
                        fields: {
                          ${this.getOptionName(fieldName)}: // corresponding field ID
                        }
                      }
                `));
        }
    }
    getOptionName(fieldName) {
        switch (fieldName) {
            case jiraIssueFetcher_1.SupportedFields.DESCRIPTION:
                return "description";
            case jiraIssueFetcher_1.SupportedFields.SUMMARY:
                return "summary";
            case jiraIssueFetcher_1.SupportedFields.LABELS:
                return "labels";
            case jiraIssueFetcher_1.SupportedFields.TEST_ENVIRONMENTS:
                return "testEnvironments";
            case jiraIssueFetcher_1.SupportedFields.TEST_PLAN:
                return "testPlan";
            case jiraIssueFetcher_1.SupportedFields.TEST_TYPE:
                return "testType";
        }
    }
}
exports.CachingJiraFieldRepository = CachingJiraFieldRepository;
