// cypress-xray-plugin/src/preprocessing/preprocessing.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCucumberPreconditionIssueTags = exports.getScenarioTagRegex = exports.getCucumberScenarioIssueTags = exports.parseFeatureFile = exports.getCucumberIssueData = exports.containsCucumberTest = exports.getNativeTestIssueKey = exports.getNativeTestIssueKeys = exports.containsNativeTest = void 0;
const gherkin_1 = require("@cucumber/gherkin");
const messages_1 = require("@cucumber/messages");
const fs_1 = __importDefault(require("fs"));
const logging_1 = require("../logging/logging");
const errors_1 = require("../util/errors");
// ============================================================================================== //
// CYPRESS NATIVE                                                                                 //
// ============================================================================================== //
function containsNativeTest(runResult, featureFileExtension) {
    return runResult.runs.some((run) => {
        if (featureFileExtension && run.spec.absolute.endsWith(featureFileExtension)) {
            return false;
        }
        return true;
    });
}
exports.containsNativeTest = containsNativeTest;
function getNativeTestIssueKeys(results, projectKey, featureFileExtension) {
    const issueKeys = [];
    for (const runResult of results.runs) {
        const keyedTests = [];
        // Cucumber tests aren't handled here. Let's skip them.
        if (featureFileExtension && runResult.spec.absolute.endsWith(featureFileExtension)) {
            continue;
        }
        for (const testResult of runResult.tests) {
            const title = testResult.title.join(" ");
            try {
                // The last element refers to an individual test (it).
                // The ones before are test suite titles (describe, context, ...).
                const issueKey = getNativeTestIssueKey(testResult.title[testResult.title.length - 1], projectKey);
                keyedTests.push(testResult);
                issueKeys.push(issueKey);
            }
            catch (error) {
                (0, logging_1.logWarning)(`Skipping test: ${title}\n\n${(0, errors_1.errorMessage)(error)}`);
            }
        }
        runResult.tests = keyedTests;
    }
    return issueKeys;
}
exports.getNativeTestIssueKeys = getNativeTestIssueKeys;
/**
 * Extracts a Jira issue key from a native Cypress test title, based on the provided project key.
 *
 * @param title - the test title
 * @param projectKey - the Jira projectk key
 * @returns the Jira issue key
 * @throws if the title contains zero or more than one issue key
 */
function getNativeTestIssueKey(title, projectKey) {
    const regex = new RegExp(`(${projectKey}-\\d+)`, "g");
    const matches = title.match(regex);
    if (!matches) {
        throw (0, errors_1.missingTestKeyInNativeTestTitleError)(title, projectKey);
    }
    else if (matches.length === 1) {
        return matches[0];
    }
    else {
        throw (0, errors_1.multipleTestKeysInNativeTestTitleError)(title, matches);
    }
}
exports.getNativeTestIssueKey = getNativeTestIssueKey;
// ============================================================================================== //
// CUCUMBER                                                                                       //
// ============================================================================================== //
function containsCucumberTest(runResult, featureFileExtension) {
    return runResult.runs.some((run) => {
        if (featureFileExtension && run.spec.absolute.endsWith(featureFileExtension)) {
            return true;
        }
        return false;
    });
}
exports.containsCucumberTest = containsCucumberTest;
function getCucumberIssueData(filePath, projectKey, isCloudClient) {
    const featureFileIssueKeys = {
        tests: [],
        preconditions: [],
    };
    const document = parseFeatureFile(filePath);
    if (!document.feature?.children) {
        return featureFileIssueKeys;
    }
    for (const child of document.feature.children) {
        if (child.scenario) {
            const issueKeys = getCucumberScenarioIssueTags(child.scenario, projectKey, isCloudClient);
            if (issueKeys.length === 0) {
                throw (0, errors_1.missingTestKeyInCucumberScenarioError)(child.scenario, projectKey, isCloudClient);
            }
            else if (issueKeys.length > 1) {
                throw (0, errors_1.multipleTestKeysInCucumberScenarioError)(child.scenario, child.scenario.tags, issueKeys, isCloudClient);
            }
            featureFileIssueKeys.tests.push({
                key: issueKeys[0],
                summary: child.scenario.name ? child.scenario.name : "<empty>",
                tags: child.scenario.tags.map((tag) => tag.name.replace("@", "")),
            });
        }
        else if (child.background) {
            const preconditionKeys = getCucumberPreconditionIssueTags(child.background, projectKey, isCloudClient, document.comments);
            if (preconditionKeys.length === 0) {
                throw (0, errors_1.missingPreconditionKeyInCucumberBackgroundError)(child.background, projectKey, isCloudClient);
            }
            else if (preconditionKeys.length > 1) {
                throw (0, errors_1.multiplePreconditionKeysInCucumberBackgroundError)(child.background, preconditionKeys, document.comments, isCloudClient);
            }
            featureFileIssueKeys.preconditions.push({
                key: preconditionKeys[0],
                summary: child.background.name ? child.background.name : "<empty>",
            });
        }
    }
    return featureFileIssueKeys;
}
exports.getCucumberIssueData = getCucumberIssueData;
/**
 * Parses a Gherkin document (feature file) and returns the information contained within.
 *
 * @param file - the path to the feature file
 * @param encoding - the file's encoding
 * @returns an object containing the data of the feature file
 * @example
 *   const data = parseFeatureFile("myTetest.feature")
 *   console.log(data.feature.children[0].scenario); // steps, name, ...
 * @see https://github.com/cucumber/messages/blob/main/javascript/src/messages.ts
 */
function parseFeatureFile(file, encoding = "utf-8") {
    const uuidFn = messages_1.IdGenerator.uuid();
    const builder = new gherkin_1.AstBuilder(uuidFn);
    const matcher = new gherkin_1.GherkinClassicTokenMatcher();
    const parser = new gherkin_1.Parser(builder, matcher);
    return parser.parse(fs_1.default.readFileSync(file, { encoding: encoding }));
}
exports.parseFeatureFile = parseFeatureFile;
function getCucumberScenarioIssueTags(scenario, projectKey, isCloudClient) {
    const issueKeys = [];
    for (const tag of scenario.tags) {
        const matches = tag.name.match(getScenarioTagRegex(projectKey, isCloudClient));
        if (!matches) {
            continue;
        }
        else if (matches.length === 2) {
            issueKeys.push(matches[1]);
        }
    }
    return issueKeys;
}
exports.getCucumberScenarioIssueTags = getCucumberScenarioIssueTags;
function getScenarioTagRegex(projectKey, isCloudClient) {
    if (isCloudClient) {
        // @TestName:CYP-123
        return new RegExp(`@TestName:(${projectKey}-\\d+)`);
    }
    // @CYP-123
    return new RegExp(`@(${projectKey}-\\d+)`);
}
exports.getScenarioTagRegex = getScenarioTagRegex;
function getCucumberPreconditionIssueTags(background, projectKey, isCloudClient, comments) {
    const preconditionKeys = [];
    if (background.steps.length > 0) {
        const backgroundLine = background.location.line;
        const firstStepLine = background.steps[0].location.line;
        for (const comment of comments) {
            if (comment.location.line > backgroundLine && comment.location.line < firstStepLine) {
                const matches = comment.text.match(backgroundRegex(projectKey, isCloudClient));
                if (!matches) {
                    continue;
                }
                else if (matches.length === 2) {
                    preconditionKeys.push(matches[1]);
                }
            }
        }
    }
    return preconditionKeys;
}
exports.getCucumberPreconditionIssueTags = getCucumberPreconditionIssueTags;
function backgroundRegex(projectKey, isCloudClient) {
    if (isCloudClient) {
        // @Precondition:CYP-111
        return new RegExp(`@Precondition:(${projectKey}-\\d+)`);
    }
    // @CYP-111
    return new RegExp(`@(${projectKey}-\\d+)`);
}
